### DISCUSSION TITLE
Help needed - Cannot use asyncio on Nucelo L432KC board?

### DISCUSSION THREAD

**Discussion by gilmanb1** [micropython/micropython#16343] [STM32 / Pyboard]
Hello gurus :) 

   I have an STM32L432KC Nucleo board and have flashed it with the latest firmware found here: https://micropython.org/download/NUCLEO_L432KC/

   I have been able to use mpremote to upload and run the code here:  /examples/hwapi/soft_pwn.py however, when I try to use the asyncio library, I have encountered a ton of pain ;). Here's what I "know":

```
Using the repl: 

>>> help('modules')
__main__          errno             math              select
_asyncio          framebuf          micropython       struct
array             gc                os                sys
binascii          hashlib           platform          time
builtins          io                pyb               uctypes
collections       json              random            vfs
deflate           machine           re
Plus any modules on the filesystem
```

I see _asyncio above but when I try: 

```
>>> import asyncio
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: no module named 'asyncio'

```

Now, I'm lost... So I've tried this:

```
>>> import sys
>>> print(sys.path)
['']
```

OK... Now I'm lost... How do I access the asyncio library? 

Thanks in advance!

**Comment by sophiedegran** [comment]
Hi
Look at your help('modules')
Seem you have to import _asyncio

**Comment by scruss** [comment]
The Nucleo L432KC default image doesn't include many standard modules because it has limited flash storage. The configuration option is here: [micropython/ports/stm32/boards/NUCLEO_L432KC/mpconfigboard.mk at 2e796d6c3e33e2d3b1936cd05aa926400ad1d5b7 · micropython/micropython](https://github.com/micropython/micropython/blob/2e796d6c3e33e2d3b1936cd05aa926400ad1d5b7/ports/stm32/boards/NUCLEO_L432KC/mpconfigboard.mk#L13)

**Reply by gilmanb1** [reply]
Thanks for the quick reply - are you saying I need to rebuild the firmware with that flag set to… what? 1 true? Thanks in advance and thanks for your patience with my novice questions here.

**Reply by scruss** [reply]
There may not be enough flash storage on the board to include all of asyncio: I couldn't see how much storage the L432KC has (256K total?).

I haven't rebuilt MicroPython in a long time, but maybe change
```
FROZEN_MANIFEST ?=
```
to
```
FROZEN_MANIFEST ?= $(BOARD_DIR)/manifest.py
```
and create a `manifest.py` in the boards/NUCLEO_L432KC/ folder containing one line:
```
include("$(PORT_DIR)/boards/manifest.py")
```
This will (? I hope) bring in [ports/stm32/boards/manifest.py](https://github.com/micropython/micropython/blob/2e796d6c3e33e2d3b1936cd05aa926400ad1d5b7/ports/stm32/boards/manifest.py), which should enable `asyncio` along with `dht` and `onewire`

**Reply by gilmanb1** [reply]
Thanks so much!! Correct 256kb flash. QQ: why does it have the _asyncio then? It looks like a private module no? Shouldn’t I be able to reference that?

**Reply by mattytrentini** [reply]
The `asyncio` module has two parts; a small core module written in C (written for efficiency) with the rest written in Python. It seems like you have the small core C module - referenced with `_asyncio` - but not the Python component.

**Reply by scruss** [reply]
So can the Python components be uploaded, assuming there's enough flash storage?

**Reply by mattytrentini** [reply]
Yes, I think so 👍🏻 I haven't tried it, but if the contents of [extmod/asyncio](https://github.com/micropython/micropython/tree/master/extmod/asyncio) are copied somewhere on the path of the device then I can't see why it wouldn't work... (?)
